### PR TITLE
Support request cache on frozen tier

### PR DESCRIPTION
**Pull request by ywelsch**
The indices request cache is currently invalidated on the frozen tier after any search, making it completely useless. This PR fixes the situation by tying the lifecycle of the cache entry to the lifecycle of the shard instead of the underlying reader.

Replaces #75309

### CODE REVIEW

**Comment by ywelsch** [comment]
CI failure is unrelated to this PR (also happened on master), will merge latest master once it's fixed.

File: server/src/main/java/org/elasticsearch/common/lucene/index/ElasticsearchDirectoryReader.java
Comment (jpountz) [inline server/src/main/java/org/elasticsearch/common/lucene/index/ElasticsearchDirectoryReader.java:70]: Can you add javadocs?

File: server/src/main/java/org/elasticsearch/common/lucene/index/ElasticsearchDirectoryReader.java
Comment (jpountz) [inline server/src/main/java/org/elasticsearch/common/lucene/index/ElasticsearchDirectoryReader.java:80]: Maybe add a comment about the `esReader == null` case. My understanding is that it should never be null, but the leniency might be ok since it "just" means that the cache would be useless, and correctness would not be affected? Maybe we still want to assert about this case?

File: x-pack/plugin/core/src/main/java/org/elasticsearch/index/engine/frozen/FrozenEngine.java
Comment (jpountz) [inline x-pack/plugin/core/src/main/java/org/elasticsearch/index/engine/frozen/FrozenEngine.java:148]: can you fix this comment to no longer reference the cache key since we check the readers directly?

File: x-pack/plugin/core/src/main/java/org/elasticsearch/index/engine/frozen/FrozenEngine.java
Comment (jpountz) [inline x-pack/plugin/core/src/main/java/org/elasticsearch/index/engine/frozen/FrozenEngine.java:290]: why are we synchonizing here?

File: x-pack/plugin/core/src/main/java/org/elasticsearch/index/engine/frozen/FrozenEngine.java
Comment (jpountz) [inline x-pack/plugin/core/src/main/java/org/elasticsearch/index/engine/frozen/FrozenEngine.java:292]: Let's validate that closed listeners are not null when registering them instead of ignoring nulls here?

File: x-pack/plugin/core/src/main/java/org/elasticsearch/index/engine/frozen/FrozenEngine.java
Comment (jpountz) [inline x-pack/plugin/core/src/main/java/org/elasticsearch/index/engine/frozen/FrozenEngine.java:143]: I'm unclear as to why we now need to have the reader itself rather than the cache key here?

**Review by jpountz — COMMENTED** [review]
I left some questions because I want to be careful given how often closed listeners have been a source of memory leaks in the past, but I like the approach, I think it's promising.

File: x-pack/plugin/core/src/main/java/org/elasticsearch/index/engine/frozen/FrozenEngine.java
Comment (ywelsch) [reply x-pack/plugin/core/src/main/java/org/elasticsearch/index/engine/frozen/FrozenEngine.java]: This is from an older iteration of the fix (where I found it easier to rewrite the code in terms of the reader than the cache key). I've reverted this part in f891f88ff1d

File: x-pack/plugin/core/src/main/java/org/elasticsearch/index/engine/frozen/FrozenEngine.java
Comment (ywelsch) [reply x-pack/plugin/core/src/main/java/org/elasticsearch/index/engine/frozen/FrozenEngine.java]: revert in f891f88ff1d

File: server/src/main/java/org/elasticsearch/common/lucene/index/ElasticsearchDirectoryReader.java
Comment (ywelsch) [reply server/src/main/java/org/elasticsearch/common/lucene/index/ElasticsearchDirectoryReader.java:92]: It should indeed never be null. I've added an assertion to that effect (cada8802697), but kept things still a bit more lenient, as this wouldn't affect correctness).

File: server/src/main/java/org/elasticsearch/common/lucene/index/ElasticsearchDirectoryReader.java
Comment (ywelsch) [reply server/src/main/java/org/elasticsearch/common/lucene/index/ElasticsearchDirectoryReader.java:79]: sure (cada8802697)

File: x-pack/plugin/core/src/main/java/org/elasticsearch/index/engine/frozen/FrozenEngine.java
Comment (ywelsch) [reply x-pack/plugin/core/src/main/java/org/elasticsearch/index/engine/frozen/FrozenEngine.java:291]: done in eb04bdf7998
I had copied the above from Lucene's way of handling `readerClosedListeners` in `SegmentReader`.

File: x-pack/plugin/core/src/main/java/org/elasticsearch/index/engine/frozen/FrozenEngine.java
Comment (ywelsch) [reply x-pack/plugin/core/src/main/java/org/elasticsearch/index/engine/frozen/FrozenEngine.java:289]: As this is `closeNoLock`, i.e., it can be called multiple times (and in parallel).

File: x-pack/plugin/searchable-snapshots/src/internalClusterTest/java/org/elasticsearch/xpack/searchablesnapshots/FrozenSearchableSnapshotsIntegTests.java
Comment (jpountz) [inline x-pack/plugin/searchable-snapshots/src/internalClusterTest/java/org/elasticsearch/xpack/searchablesnapshots/FrozenSearchableSnapshotsIntegTests.java:558]: Can you also test that the cache entries are removed when closing the shard (via the number of evictions or the memory size)?

**Review by jpountz — APPROVED** [review]
I like the approach and the simplifications from recent commits helped me get more confident that it is correct. I'd just like @jimczi to have a quick look at this before merging.

**Review by jimczi — APPROVED** [review]
LGTM too

File: x-pack/plugin/searchable-snapshots/src/internalClusterTest/java/org/elasticsearch/xpack/searchablesnapshots/FrozenSearchableSnapshotsIntegTests.java
Comment (ywelsch) [reply x-pack/plugin/searchable-snapshots/src/internalClusterTest/java/org/elasticsearch/xpack/searchablesnapshots/FrozenSearchableSnapshotsIntegTests.java:573]: I wanted to test that, but couldn't come up with a good way to do it without reaching directly into the implementation of IndicesRequestCache. The reason it's hard is that the stats are tracked on a per-shard level (within IndexShard), and so vanish when the shard is shut down. I've added some basic infra to reach into the IndicesRequestCache directly and extract the relevant information in ea607eb2632.

**Comment by jpountz** [comment]
@ywelsch FYI I added the `release highlight`label.

**Pull request merged**
